src/Interpreters/TreeRewriter.cpp

```diff
@@ -693,7 +886,7 @@ void TreeRewriterResult::collectUsedColumns(const ASTPtr & query, bool is_select
             if (required.count(name))
             {
                 /// Optimisation: do not add columns needed only in JOIN ON section.
-                if (columns_context.nameInclusion(name) > analyzed_join->rightKeyInclusion(name))
+                if (columns_context.converted_to_dnf || columns_context.nameInclusion(name) > analyzed_join->rightKeyInclusion(name))
                     analyzed_join->addJoinedColumn(joined_column);
```

Comment:
It is not the same (a OR b is not converted_to_dnf),
but I agree that it looks strange and I have to revise this logic.

Reply:
Ok, the original logic is
_calculate how many times a column occurred in a query, compare with how many times it occurred in join, if first is grater than second we call addJoinedColumn for the column._
The problem here is DNFing applied to table_join.on_query, not to original query, so calculation of 'how many times a column occurred in a query' is not affected by DNFing, but calculation of 'how many times it occurred in join' is affected. 
So, columns_context.nameInclusion(name) > analyzed_join->rightKeyInclusion(name) does not work as expected.

In other words.
Imagine that we have a query   _select A  .. join    ON  A_    columns_context.nameInclusion(name) is 2, analyzed_join->rightKeyInclusion(name) is 1 and we call addJoinedColumn
If because of DNFing the query is logically transformed to _select A  .. join    ON  A    A _  , columns_context.nameInclusion(name) is 2, but analyzed_join->rightKeyInclusion(name) is 2 too and we do not call addJoinedColumn despite it is necessary.

Reply:
Plan to keep as is, but add a comment.